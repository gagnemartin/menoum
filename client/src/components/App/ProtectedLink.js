import { useEffect, useState } from 'react';
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { useUserState } from '../../hooks/useUser'

const ProtectedLink = (props) => {
  const userData = useUserState()
  const { children, role, to } = props
  const [isLoading, setIsLoading] = useState(true)
  const [isAuthenticated, setIsAuthenticated] = useState(false)
  const [isAuthorized, setIsAuthorized] = useState(false)

  const checkAuthenthicated = () => {
    const { status } = userData

    return status === 'SUCCESS'
  }

  const checkAuthorized = () => {
    const { status, user } = userData

    return status === 'SUCCESS' && user.role === role
  }

  useEffect(() => {
    if (userData.status === 'REQUEST') {
      setIsLoading(true)
    } else {
      setIsAuthenticated(checkAuthenthicated())
      setIsAuthorized(checkAuthorized())
      setIsLoading(false)
    }
  }, [userData])

  if (isLoading) return null

  if (isAuthenticated && isAuthorized) {
    return <Link to={to}>{children}</Link>
  }

  return null
}

ProtectedLink.propTypes = {
  to: PropTypes.string.isRequired,
  role: PropTypes.string
}

ProtectedLink.defaultProps = {
  role: 'user'
}

export default ProtectedLink
