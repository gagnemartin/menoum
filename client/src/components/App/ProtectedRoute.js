import { useState, useEffect } from 'react';
import PropTypes from 'prop-types'
import { Redirect, Route } from 'react-router-dom'
import { useUserState } from '../../hooks/useUser'

const ProtectedRoute = (props) => {
  const { children, component, exact, path, role } = props
  const userData = useUserState()
  const [isLoading, setIsLoading] = useState(true)
  const [isAuthenticated, setIsAuthenticated] = useState(false)
  const [isAuthorized, setIsAuthorized] = useState(false)
  const Component = component

  const checkAuthenthicated = () => {
    const { status } = userData

    return status === 'SUCCESS'
  }

  const checkAuthorized = () => {
    const { status, user } = userData

    return status === 'SUCCESS' && user.role === role
  }

  useEffect(() => {
    if (userData.status === 'REQUEST') {
      setIsLoading(true)
    } else {
      setIsAuthenticated(checkAuthenthicated())
      setIsAuthorized(checkAuthorized())
      setIsLoading(false)
    }
  }, [userData])

  if (isLoading) return null

  if (isAuthenticated) {
    if (!isAuthorized) {
      return <Redirect to={{ pathname: '/' }} />
    }
    
    if (component) {
      return <Component />
    }

    return (
      <Route exact={exact} path={path}>
        {children}
      </Route>
    )
  } else {
    return <Redirect to={{ pathname: '/login' }} />
  }
}

ProtectedRoute.propTypes = {
  exact: PropTypes.bool,
  path: PropTypes.string.isRequired,
  role: PropTypes.string
}

ProtectedRoute.defaultProps = {
  exact: false,
  role: 'user'
}

export default ProtectedRoute
