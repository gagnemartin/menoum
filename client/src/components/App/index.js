import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
import { Home, NewRecipe, UpdateRecipe, Login, Register } from '../../pages'
import { UserProvider } from '../../context/userContext'
import { ProtectedRoute } from '../Protected'
import Navigation from '../Navigation'

const App = () => {
  return (
    <UserProvider>
      <Router>
        <div>
          <Navigation />

          <Switch>
            <ProtectedRoute role='admin' exact={true} path='/recipe/new'>
              <NewRecipe />
            </ProtectedRoute>

            <ProtectedRoute role='admin' exact={true} path='/recipe/edit/:uuid'>
              <UpdateRecipe />
            </ProtectedRoute>

            <Route path='/login'>
              <Login />
            </Route>

            <Route path='/register'>
              <Register />
            </Route>

            <Route path='/'>
              <Home />
            </Route>
          </Switch>
        </div>
      </Router>
    </UserProvider>
  )
}

export default App
