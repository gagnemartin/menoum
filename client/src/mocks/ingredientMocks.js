const mockIngredients = [
  {
    uuid: 'fakeIngredientUuid',
    name: 'Potato'
  }
]

const mockIngredientsServiceResponse = {
  search: {
    success: {
      status: 'success',
      data: mockIngredients
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  }
}

export { mockIngredientsServiceResponse, mockIngredients }
