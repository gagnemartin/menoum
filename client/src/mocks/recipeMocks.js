const mockRecipes = [
  {
    id: 199,
    uuid: '2dff4911-4596-4dc9-86d4-ce640fd6aeed',
    name: 'Voluptatum dolorum qui fuga nisi.',
    steps: [
      {
        type: 'section',
        value: 'Salad',
        steps: [
          {
            type: 'step',
            value: 'Iure alias labore consequatur officia.'
          },
          {
            type: 'step',
            value: 'Labore et sed eligendi dolorem.'
          },
          {
            type: 'step',
            value: 'Molestiae cum distinctio dolorem et.'
          }
        ]
      },
      {
        type: 'step',
        value: 'Non expedita nostrum quo libero.'
      },
      {
        type: 'step',
        value: 'Soluta qui nihil pariatur qui.'
      },
      {
        type: 'step',
        value: 'Iusto sint hic voluptatem et.'
      },
      {
        type: 'step',
        value: 'Repudiandae eligendi dolorem eum quo.'
      },
      {
        type: 'step',
        value: 'Ratione aliquid possimus est soluta.'
      },
      {
        type: 'step',
        value: 'Vel et consequatur accusamus labore.'
      }
    ],
    created_at: '2021-03-27T18:57:44.951Z',
    updated_at: '2021-03-27T18:57:44.951Z',
    ingredients: [
      {
        id: 396,
        name: 'fresh tuna',
        uuid: 'ddbc2042-e807-4bf4-98fa-8217589992a9',
        elastic_id: 'D1EMdXgBdqz25pkfoIkz',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1029,
          ingredient_id: 396,
          recipe_id: 199,
          unit: 'g',
          amount: 5,
          section: 'Bread',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 964,
        name: 'summer cabbage',
        uuid: '871ea847-b9c7-4ea1-8e39-b821c69b33e1',
        elastic_id: 'R1EMdXgBdqz25pkfoIs2',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1021,
          ingredient_id: 964,
          recipe_id: 199,
          unit: 'ml',
          amount: 5,
          section: 'Starter',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 1027,
        name: 'vanilla essence',
        uuid: 'f13e30a4-08f5-446a-82c4-9eeb8cb022fd',
        elastic_id: 'hlEMdXgBdqz25pkfoIs3',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1027,
          ingredient_id: 1027,
          recipe_id: 199,
          unit: 'ml',
          amount: 4,
          section: 'Pasta',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 612,
        name: 'mint sauce',
        uuid: 'b4376bea-38c2-4249-9195-bb2db455e69b',
        elastic_id: '51EMdXgBdqz25pkfoIk1',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1026,
          ingredient_id: 612,
          recipe_id: 199,
          unit: 'ml',
          amount: 7,
          section: 'Bread',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 131,
        name: 'buttercream icing',
        uuid: '02fb5ee5-1728-4cf1-99d8-810d21816c83',
        elastic_id: 'BlEMdXgBdqz25pkfoIgx',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1025,
          ingredient_id: 131,
          recipe_id: 199,
          unit: 'g',
          amount: 10,
          section: 'Sauce',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 313,
        name: 'danish blue',
        uuid: '77fe7320-9417-449b-af72-9421da8daf96',
        elastic_id: 'vFEMdXgBdqz25pkfoIgy',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1024,
          ingredient_id: 313,
          recipe_id: 199,
          unit: 'g',
          amount: 7,
          section: 'Appetizers',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 359,
        name: 'fenugreek',
        uuid: 'fc6f2954-4cf2-4d58-8f80-cdd72bc554fa',
        elastic_id: '6lEMdXgBdqz25pkfoIgz',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1023,
          ingredient_id: 359,
          recipe_id: 199,
          unit: 'ml',
          amount: 1,
          section: 'Salad',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 667,
        name: 'orange',
        uuid: '551c5e6b-805f-45f9-b80d-7efdff1cf816',
        elastic_id: 'HlEMdXgBdqz25pkfoIo1',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1022,
          ingredient_id: 667,
          recipe_id: 199,
          unit: 'ml',
          amount: 2,
          section: 'Salad',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 1034,
        name: 'vegetables',
        uuid: '522c6c91-480a-4eb2-81e8-b37139f10e91',
        elastic_id: 'jVEMdXgBdqz25pkfoIs3',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1020,
          ingredient_id: 1034,
          recipe_id: 199,
          unit: 'ml',
          amount: 1,
          section: 'Starter',
          weight: 2.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      },
      {
        id: 700,
        name: 'paw-paw',
        uuid: 'e5ae0f5d-77e6-4169-a6c8-d42a6f6e19de',
        elastic_id: 'P1EMdXgBdqz25pkfoIo1',
        created_at: '2021-03-27T18:57:44.930658',
        updated_at: '2021-03-27T18:57:44.930658',
        recipe_count: 0,
        ingredients_recipes: {
          id: 1028,
          ingredient_id: 700,
          recipe_id: 199,
          unit: 'g',
          amount: 4,
          section: 'Pasta',
          weight: 1.0,
          created_at: '2021-03-27T18:57:47.013142',
          updated_at: '2021-03-27T18:57:47.013142'
        }
      }
    ]
  }
]

const mockRecipesServiceResponse = {
  suggest: {
    success: {
      status: 'success',
      data: mockRecipes
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  },
  add: {
    success: {
      status: 'success',
      data: mockRecipes[0]
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  },
  update: {
    success: {
      status: 'success',
      data: mockRecipes[0]
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  },
  get: {
    success: {
      status: 'success',
      data: mockRecipes[0]
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  }
}

export { mockRecipesServiceResponse, mockRecipes }
