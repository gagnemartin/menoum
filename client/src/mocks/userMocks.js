import jwt from 'jsonwebtoken'

const mockUser = {
  email: 'test@test.com',
  password: 'test123',
  role: 'user'
}

const mockUserAdmin = {
  ...mockUser,
  role: 'admin'
}

const userToken = jwt.sign(mockUser, 'fakehash')
const userAdminToken = jwt.sign(mockUserAdmin, 'fakehash')

const mockUsersServiceResponse = {
  refresh: {
    success: {
      status: 'success',
      data: {
        token: userToken
      }
    },
    successAdmin: {
      status: 'success',
      data: {
        token: userAdminToken
      }
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  },

  login: {
    success: {
      status: 'success',
      data: {
        token: userToken
      }
    },
    successAdmin: {
      status: 'success',
      data: {
        token: userAdminToken
      }
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  },
  
  logout: {
    success: {
      status: 'success',
      data: {
        token: userToken
      }
    },
    successAdmin: {
      status: 'success',
      data: {
        token: userAdminToken
      }
    },
    error: {
      status: 'error',
      error: new Error('False Error')
    }
  },

  register: {
    success: {
      status: 'success',
      data: {
        token: userToken
      }
    },
    successAdmin: {
      status: 'success',
      data: {
        token: userAdminToken
      }
    },
    error: {
      status: 'error',
      error: new Error('False Error'),
      data: {
        email: { minLength: 'Too short' }
      }
    }
  }
}

export { mockUserAdmin, mockUser, mockUsersServiceResponse, userAdminToken, userToken }
