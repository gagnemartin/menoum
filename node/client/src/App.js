import React, { useState, useEffect } from 'react';
// import logo from './logo.svg';
// import './App.css';

function App() {
  const [searchInputValue, setSearchInputValue] = useState('')
  const [suggestedIngredients, setSuggestedIngredients] = useState([])
  const [selectedIngredients, setSelectedIngredients] = useState([])

  const [suggestedRecipes, setSuggestedRecipes] = useState([])

  const handleChange = (e) => {
    const value = e.target.value
    setSearchInputValue(value)
  }

  const handleClickIngredient = (e) => {
    const uuid = e.target.dataset.uuid
    const ingredient = suggestedIngredients.find((d) => d.uuid === uuid)

    if (ingredient) {
      setSelectedIngredients([...selectedIngredients, ingredient])
      setSuggestedIngredients([])
      setSearchInputValue('')
    }
  }

  const handleClickRemove = (e) => {
    const uuid = e.target.dataset.uuid
    const ingredient = selectedIngredients.find((d) => d.uuid === uuid)

    if (ingredient) {
      setSelectedIngredients(
        selectedIngredients.filter((d) => d.uuid !== ingredient.uuid)
      )
    }
  }

  // const searchIngredients = async (q) => {
  //   const data = await fetch(`/api/ingredients/search?q=${searchInputValue}`)

  //   console.log(data)
  // }

  useEffect(() => {
    if (searchInputValue.trim().length > 0) {
      // searchIngredients(searchInputValue)
      fetch(
        `http://localhost:4000/api/v1/ingredients/search?q=${searchInputValue}`
      )
        .then((data) => {
          return data.json()
        })
        .then((data) => {
          const ingredients = data.filter((ingredient) => {
            return !selectedIngredients.some((d) => d.uuid === ingredient.uuid)
          })

          setSuggestedIngredients(ingredients)
        })
    } else {
      setSuggestedIngredients([])
    }
  }, [searchInputValue])

  useEffect(() => {
    if (selectedIngredients.length > 0) {
      const uuids = selectedIngredients.map((ingredient) => ingredient.uuid)
      const queryString = uuids
        .map((uuid) => {
          return `uuids[]=${uuid}`
        })
        .join('&')

      fetch(`http://localhost:4000/api/v1/recipes/suggest?${queryString}`)
        .then((data) => {
          return data.json()
        })
        .then((data) => {
          setSuggestedRecipes(data)
        })
    } else {
      setSuggestedRecipes([])
    }
  }, [selectedIngredients])

  return (
    <div className='App'>
      <input type='text' value={searchInputValue} onChange={handleChange} />

      <div style={{ margin: '10px' }}>
        {selectedIngredients.map((ingredient) => (
          <span
            key={ingredient.uuid}
            style={{
              border: 'solid 1px lightgray',
              marginRight: '10px',
              padding: '5px',
            }}
          >
            {ingredient.name}
            <button
              onClick={handleClickRemove}
              data-uuid={ingredient.uuid}
              style={{ marginLeft: '10px' }}
            >
              X
            </button>
          </span>
        ))}
      </div>

      <div>
        {suggestedIngredients.map((ingredient) => (
          <p
            key={ingredient.uuid}
            data-uuid={ingredient.uuid}
            onClick={handleClickIngredient}
          >
            {ingredient.name}
          </p>
        ))}
      </div>

      <div
        style={{
          display: 'grid',
          gridTemplateColumns: 'repeat(4, 1fr)',
          columnGap: '20px',
        }}
      >
        {suggestedRecipes.map((recipe) => (
          <div key={recipe.uuid}>
            <img src={recipe.thumbnail} width='100%' />
            <p>{recipe.name}</p>
          </div>
        ))}
      </div>
    </div>
  )
}

export default App;
