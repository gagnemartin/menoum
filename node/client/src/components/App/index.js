import React from 'react'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";
import { Home, NewRecipe } from '../../pages'

function App() {

  return (
    <Router>
      <div>
        <nav>
          <ul>
            <li>
              <Link to="/">Home</Link>
            </li>
            <li>
              <Link to="/recipe/new">Add a Recipe</Link>
            </li>
          </ul>
        </nav>

        <Switch>
          <Route path="/recipe/new">
            <NewRecipe />
          </Route>
          
          <Route path="/">
            <Home />
          </Route>
        </Switch>
      </div>
    </Router>
  )
}

export default App;
