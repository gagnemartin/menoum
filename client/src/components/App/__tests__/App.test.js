import { render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import App from '../'
import UsersService from '../../../services/usersService'
import { mockUsersServiceResponse } from '../../../mocks/userMocks'

jest.mock('../../../services/usersService')

describe('<App />', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  it.skip('should render and match snapshot', async () => {
    UsersService.refresh.mockReturnValue(mockUsersServiceResponse.refresh.success)
    const { asFragment } = render(<App />)

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })

  it('should navigate to admin section for admin user', async () => {
    UsersService.refresh.mockReturnValue(mockUsersServiceResponse.refresh.successAdmin)
    const { queryByTestId, queryByText } = render(<App />)
    let elementEmail
    let element

    await waitFor(() => {
      elementEmail = queryByTestId('nav-email')
      expect(elementEmail).toBeInTheDocument()
    })

    userEvent.click(elementEmail)

    await waitFor(() => {
      element = queryByTestId('nav-new-recipe')
    })

    userEvent.click(element)

    await waitFor(() => {
      const title = queryByText(/New Recipe/)
      expect(title).toBeInTheDocument()
    })
  })
})
