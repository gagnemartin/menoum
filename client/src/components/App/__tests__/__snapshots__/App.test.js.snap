// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render and match snapshot 1`] = `
<DocumentFragment>
  <div>
    <nav>
      <div
        class="MuiBox-root css-k008qs"
      >
        <div
          class="MuiBox-root css-1ilyui9"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-tgjvbr-MuiTypography-root-MuiLink-root"
            data-testid="nav-home"
            href="/"
          >
            Home
          </a>
        </div>
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            data-testid="nav-loading"
          >
            loading
          </p>
        </div>
      </div>
    </nav>
    <div>
      <input
        data-testid="ingredient-search-input"
        type="text"
        value=""
      />
      <div
        style="position: relative;"
      >
        <div
          style="position: absolute; top: 0px; left: 0px; padding: 10px; background: rgb(255, 255, 255); display: none;"
        />
      </div>
      <div
        style="margin: 10px;"
      />
    </div>
    <div
      style="display: grid; grid-template-columns: repeat(4, 1fr); column-gap: 20px;"
    />
  </div>
</DocumentFragment>
`;
